langchain/chains/tot/base.py

```diff
@@ -0,0 +1,317 @@
+        arbitrary_types_allowed = True
+
+    @property
+    def tot_prompter(self) -> ToTPrompter:
+        return ToTPrompter(
+            first_step_prompt=self.first_step_prompt,
+            next_step_prompt=self.next_step_prompt,
+        )
+
+    @property
+    def input_keys(self) -> List[str]:
+        """Will be whatever keys the prompt expects.
+
+        :meta private:
+        """
+        return self.first_step_prompt.input_variables
+
+    @property
+    def output_keys(self) -> List[str]:
+        """Will always return text key.
+
+        :meta private:
+        """
+        return [self.output_key]
+
+    def parse_llm_output(self, llm_output: str) -> str:
+        """Parse the output of the language model."""
+        if match := re.search(r"\{.*?\}", llm_output, re.DOTALL):
+            try:
+                return json.loads(match.group())["next_step"]
+            except json.JSONDecodeError:
+                return ""
+        return ""
+
+    @property
+    def current_level(self) -> int:
+        return len(self.tot_memory.stack)
+
+    def log_request(
+        self, prompt: str, run_manager: Optional[CallbackManagerForChainRun] = None
+    ) -> None:
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Request >>>\n{prompt}\n\n", prefix),
+                verbose=self.verbose,
+                color="green",
+            )
+
+    def log_response(
+        self, response: str, run_manager: Optional[CallbackManagerForChainRun] = None
+    ) -> None:
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Response >>>\n{response}\n\n", prefix),
+                verbose=self.verbose,
+                color="green",
+            )
+
+    def log_ctrl_signal(
+        self,
+        ctrl_signal: Optional[str],
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> None:
+        if ctrl_signal is None:
+            ctrl_signal = "<None>"
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Ctrl Signal >>>\n{ctrl_signal}\n\n", prefix),
+                verbose=self.verbose,
+                color="red",
+            )
+
+    def solve(
+        self,
+        problem_description: str,
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> Optional[str]:
+        """Algorithm 2 from the ToT paper."""
+        checker = partial(self.checker, problem_description)
+        llm = self.llm.predict
+        controller = self.tot_controller
+        prompter = self.tot_prompter
+        memory = self.tot_memory
+
+        prompt = prompter(problem_description)
+        for _ in range(self.k):
+            self.log_request(prompt, run_manager)
+            response = self.parse_llm_output(llm(prompt))
```

Comment:
we should construct and use LLMChain here - will help with logging (may make log_request and log_response unneccessary